data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java

```diff
@@ -24,15 +42,96 @@
 public class KafkaSource implements Source<Record<Object>> {
 
   private static final Logger LOG = LoggerFactory.getLogger(KafkaSource.class);
-  
+  private final KafkaSourceConfig sourceConfig;
+  private Set<MultithreadedConsumer> consumers = new HashSet<>();
+  private ExecutorService executorService;
+  private static final String KAFKA_WORKER_THREAD_PROCESSING_ERRORS = "kafkaWorkerThreadProcessingErrors";
+  private final Counter kafkaWorkerThreadProcessingErrors;
+  private final PluginMetrics pluginMetrics;
+  private String consumerGroup;
+  private MultithreadedConsumer multithreadedConsumer;
+  private int totalWorkers;
+
+  @DataPrepperPluginConstructor
+  public KafkaSource(final KafkaSourceConfig sourceConfig, final PluginMetrics pluginMetrics) {
+    this.sourceConfig = sourceConfig;
+    this.pluginMetrics = pluginMetrics;
+    this.kafkaWorkerThreadProcessingErrors = pluginMetrics.counter(KAFKA_WORKER_THREAD_PROCESSING_ERRORS);
+  }
+
   @Override
   public void start(Buffer<Record<Object>> buffer) {
-    LOG.info("Initiated the start() method...");
+    sourceConfig.getTopics().forEach(topic -> {
+      totalWorkers = 0;
+      Properties consumerProperties = getConsumerProperties(topic);
+      try {
+        totalWorkers = topic.getTopic().getConsumerGroupConfig().getWorkers();
+        consumerGroup = topic.getTopic().getConsumerGroupConfig().getGroupName();
+        executorService = Executors.newFixedThreadPool(totalWorkers);
+        IntStream.range(0, totalWorkers).forEach(index -> {
+          String consumerId = consumerGroup +"::"+ topic.getTopic().getName()+"::"+Integer.toString(index + 1);
+          multithreadedConsumer = new MultithreadedConsumer(consumerId,
+                  consumerGroup, consumerProperties, topic.getTopic(), buffer, pluginMetrics);
+          consumers.add(multithreadedConsumer);
+          executorService.submit(multithreadedConsumer);
+        });
+      } catch (Exception e) {
+        LOG.error("Failed to setup the Kafka Source Plugin.", e);
+        kafkaWorkerThreadProcessingErrors.increment();
+      }
+    });
   }
 
   @Override
   public void stop() {
-   LOG.info("Initiated the stop method...");
+    LOG.info("Shutting down Consumers...");
+    executorService.shutdown();
+    try {
+      if (!executorService.awaitTermination(
+              sourceConfig.getTopics().get(0).getTopic().getConsumerGroupConfig().getThreadWaitingTime().toSeconds(), TimeUnit.MILLISECONDS)) {
```

Comment:
why just get(0). I think there was a comment to wait for longest wait time of all consumer threads

Reply:
resolved